fundraiser/src/main.sw

```diff
@@ -0,0 +1,502 @@
+        }
+
+        // Update the amount that they have pledged
+        pledge.amount = pledge.amount - amount;
+
+        // Lower the campaign total pledge by the amount the user has unpledged
+        campaign_info.total_pledge = campaign_info.total_pledge - amount;
+
+        // Update the state of their pledge with the new version
+        storage.pledge_history.insert((user, pledge_history_index), pledge);
+
+        // Update the campaign state with the updated version as well
+        storage.campaign_info.insert(id, campaign_info);
+
+        // Update the asset amount to track the removal of the amount
+        let mut asset_info = storage.asset_info.get(campaign_info.asset);
+        asset_info.amount = asset_info.amount - amount;
+
+        // Update asset state
+        storage.asset_info.insert(campaign_info.asset, asset_info);
+
+        // Transfer back the amount the user has unpledged
+        transfer(amount, campaign_info.asset, user);
+
+        // We have updated the state of a campaign therefore we must log it
+        log(UnpledgedEvent {
+            amount, id, user
+        });
+    }
+
+    /// Returns the total number of campaigns that have been created by all users
+    #[storage(read)]fn total_campaigns() -> u64 {
+        storage.total_campaigns
+    }
+
+    /// Returns information about the specified campaign
+    ///
+    /// # Arguments
+    ///
+    /// * `id` - Unique campaign identifier which is a number from the storage.total_campaigns range
+    ///
+    /// # Reverts
+    ///
+    /// * When the `id` is either 0 or greater than the total number of campaigns created
+    #[storage(read)]fn campaign_info(id: u64) -> CampaignInfo {
+        // User cannot interact with a non-existent campaign
+        validate_id(id, storage.total_campaigns);
+        storage.campaign_info.get(id)
+    }
+
+    /// Returns the number of campaigns that the user has created
+    ///
+    /// # Reverts
+    ///
+    /// * When an AuthError is generated
+    #[storage(read)]fn user_campaign_count() -> u64 {
+        storage.user_campaign_count.get(sender_identity())
+    }
+
+    /// Returns information about the specified campaign for the campaign author
+    ///
+    /// # Arguments
+    ///
+    /// * `id` - Unique identifier which is a number starting from 1...storage.user_campaign_count
+    ///
+    /// # Reverts
+    ///
+    /// * When the `id` is either 0 or greater than the total number of campaigns created by the author
+    /// * When an AuthError is generated
+    #[storage(read)]fn campaign(id: u64) -> Campaign {
+        // Validate the ID to ensure that the user has created the campaign
+        validate_id(id, storage.user_campaign_count.get(sender_identity()));
+        storage.campaign_history.get((sender_identity(), id))
+    }
+
+    /// Returns the number of campaigns that the user has pledged to
+    ///
+    /// # Reverts
+    ///
+    /// * When an AuthError is generated
```

Comment:
I'm not sure what the correct approach is here but do we need to list all possible reverts all the way down the stack, or just in the specific function we're documenting?
AutError here is from the stdlib, and probably doesn't need to be mentioned in every function which uses `msg_sender()`?

Reply:
I should only be documenting errors generated by the contract and not the library itself. I'll make all of these adjustments soon